feat: add progress bar to the lower jaw

Ismailtlem opened the pull request:
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to #46178

ahmadabdolsaheb commented:
Thanks for working on this feature, @Ismailtlem. 
I think @JordanMooree based on this [comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/46429#issuecomment-1167345263). I am mentioning him here so you could work on this feature together if you wish. 

Notes:
It would be great if you could isolate the progress bar that is used in the modal so it only receives a couple of props and renders a progress bar. That way we code would more modular.

Ismailtlem commented:
Hi @JordanMooree 
Yes, please go ahead.
I will write the progress and issues I faced here asap

Ismailtlem commented:
Hi,

- The first problem I had here is how to get the challenges data to get the percentage on the progress bar. I get the challenges data of each block from this component https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Introduction/components/challenges.tsx by storing it on local storage and  then getting it on https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Challenges/classic/lower-jaw.tsx. It works if you get to the challenges block first and then access a project. There is still an edge case where we have an issue : if you clear cache and directly access a step directly by its link. There must be some way to lift this data up and then get it back on the lower jaw component I guess or some better way
- There is the test that needs to be fixed.
- I can make the progress bar into a separate component

ahmadabdolsaheb commented:
I might be mistaken, but doesn't the logic that works in the completion modal to show percentages also work here?
In that case, you would not need to re create the logic.

Turning the completion modal's progress bar into its own component and passing it to the modal and the editor lower jaw should do that trick. 

let me know if that makes sense.

Ismailtlem commented:
Hi @ahmadabdolsaheb sorry for my late answer.
I think I understood what you said, but the logic of the percentage shown on the progress bar is in the [completion modal body](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Challenges/components/completion-modal-body.tsx) which is not rendered in the new editor. Shouldn't we recreate the logic of the percentage shown on the progress bar ?

ahmadabdolsaheb commented:
Anything inside the `progress-bar-wrap` should be extracted into its own component so it would receive a couple of props and display the progress bar animation.  That way we could use the new component in the modal and pass it to the lower jaw in the editor. If you need more clarification, contact me through chat so we can hop on a call.

JordanMooree commented:
> Anything inside the progress-bar-wrap should be extracted into its own component so it would receive a couple of props and display the progress bar animation.

@ahmadabdolsaheb Are you looking for a similar animation like done in the Legacy RWD?
![progress-animation ](https://user-images.githubusercontent.com/27147016/184231736-90e68f2d-8ac0-4081-8ae7-34af5b30098e.gif)

ahmadabdolsaheb commented:
@JordanMooree yes.

Ismailtlem commented:
Hi,

I think I am getting close. I have a progress bar displaying the percentage of the completed challenges. There is still some work to do on the code.
![Screenshot from 2022-08-16 00-33-31](https://user-images.githubusercontent.com/34961373/184755516-659bc890-1907-4aa9-af37-906f2b1d2a78.png)

- We need to add the animation on the progress bar to show the percentage gradually like in the legacy certification. Since [the code](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Challenges/components/completion-modal-body.tsx) responsible for the shown percentage will be the same I guess in the new editor, we should maybe put it somewhere and import it in the new editor and in the legacy challenges.
- We still need to fix the linting errors
- I am not sure if the naming "blockCompletedPercent" of the completed percentage on the lower jaw component is good enough since the name "completedPercent" was already used for the legacy challenges.

Sembauke reviewed (dismissed):
One note, I am not sure if it should always be visible. This should be consistent with the old legacy RWD in my opinion.

ahmadabdolsaheb commented:
Thank you for the hard work. It is definitely coming together.
Here are a few things to do before this PR is finalized:

1. The progress bar should show up after the tests has passed.
2. The animation related logic should be relocated inside the progress bar.
3. The completion modal’s progress bar should be replaced with the new progress bar.
4. Remove disabled eslint rules 
5. Passing around `blockCompletedPercent` and  `setAttemptedInNewEditor` should be avoided.
6. The percentage calculations should be relocated to ‘Show’ and passed to `Multifile-editor` and `lower jaw`
7. Write tests using cypress.

Note:
This is out of the scope for this pr, but we need to move the percentage and percentage calculations to redux, so it could be accessed from the multifile-editor` and `lower jaw` and the donation-saga ( for opening the block donation modals)
Also the graphql query should be optimized so we don’t filter after receiving the data.

Ismailtlem commented:
Hi @ahmadabdolsaheb 

I went through the points you mentioned. Let me know if I missed something.

There is still a warning raised by CodeFactor we need to fix (conflicts with what eslint suggests).

moT01 reviewed (changes requested):
This is looking good @Ismailtlem - I took a look and found a few things though...

- the progress bar element is slightly taller than the restart button right below it. Perhaps that's fine since it's not a button.

- Some campers may be confused what the progress bar is for. I suppose they should pick up on it pretty quick - but we show the block name in the completion modal to denote what it's for. Here we just show the percent. I'm fine with that - just mentioning it.

- Depending on where I am scrolled to, the lower jaw is sometimes off screen. This is probably cause I was copy/pasting solutions - we could scroll to that area when the tests pass? Maybe another discussion/PR.

^ These last two things made me wonder why we don't just use the completion modal for these @ahmadabdolsaheb? It would unify some of the experience. We already have it ready, and it isn't any more clicks to get to the next step.

- Something is slightly off in the calculations. I finished through step 6 of the photo gallery project and this progress bar showed 29%, but the main map bar showed 28% - saw the same thing on other steps. They were the same on many of the steps though. Maybe we should extract how that is calculated and use the same function for both areas.

- This is the only problem that we should probably fix here:
![Screen Shot 2022-09-28 at 5 22 39 PM](https://user-images.githubusercontent.com/20648924/192902238-7caef004-8ef2-4030-ab3e-562271e0041f.png)

![Screen Shot 2022-09-28 at 5 32 22 PM](https://user-images.githubusercontent.com/20648924/192902250-d01462af-e029-4a02-a5f0-50a9b87aaa7e.png)
^ the text is off center. Saw it on firefox and chrome.

Regarding the codefactor issues - we may be able to just ignore them. That only runs on PR's I think so the tests wouldn't fail on main. Otherwise, can you not add a `eslint-disable testing-library/no-unnecessary-act` or something?

ahmadabdolsaheb commented:
Thank you @Ismailtlem for refactoring the components and adding the progress bar to the lowerjaw. The pull request looks almost ready for a merge.

Here are my thoughts:
- If we are only using the current block Ids to calculate the `completedPercent`, you should query `currentBlockIds` as high up the component hierarchy, calculate `completedPercent`, and pass `completedPercent` props to children.  That would simplify the flow of information. If show is the parent component for two of the consumers of the `completedPercent`, let's calculate `completedPercent` there once and pass it to `MultifileEditor` and `completion-modal`.
- Avoid styling components in JS. You could style components based on their parents, or add a default style with a conditional className.
- The timeout in get in the cypress test should be fine as it the test proceeds as soon as the element is found, but I am not sure about the completion-modal-body one. I would appreciate it if you could rewrite the completion-modal-body test to avoid slowing down the CI. 
- Some es-lint rules are disabled in the`show` component. It would be great if you could modify the components to follow es-lint rules if possible.

Thanks for the review @moT01 and the feedback.
> Some campers may be confused what the progress bar is for. I suppose they should pick up on it pretty quick - but we show the block name in the completion modal to denote what it's for. Here we just show the percent. I'm fine with that - just mentioning it.

That is a valid point. Maybe we should mention "Course" in that line somewhere to prevent confusion for now.

>Depending on where I am scrolled to, the lower jaw is sometimes off screen. This is probably cause I was copy/pasting solutions - we could scroll to that area when the tests pass? Maybe another discussion/PR.

That is also a good point. Once we turn Help and Restart buttons into icons, we could explore having those buttons and the progress bar in the same row in an attempt to make the lower jaw shorter.

>^ These last two things made me wonder why we don't just use the completion modal for these @ahmadabdolsaheb? It would unify some of the experience. We already have it ready, and it isn't any more clicks to get to the next step.
Having everything in the lower jaw is an attempt to lower the friction, but I don't have any hard data yet.

General note: 
The progress bar that close to the button make it look like a button. Once this pr is ready to merge, I will go in and change the styles slightly to prevent confusion.

moT01 commented:
I see a lot of tests failing here so I put the waiting update label back on here. Let us know when you are ready for another round of reviews @Ismailtlem 👍

bbsmooth commented on client/src/components/ProgressBar/progress-bar.tsx:
```diff
@@ -0,0 +1,98 @@
+  return (
+    <div
+      className='progress-bar-wrap'
+      aria-label={t('learn.percent-complete', {
+        percent: completedPercent
+      })}
```
```suggestion
```
The JAWS screen reader does not respect `aria-label` on the `div` element. So we'll need to add this text using sr-only text instead.

bbsmooth commented on client/src/components/ProgressBar/progress-bar.tsx:
```diff
@@ -0,0 +1,98 @@
+      })}
+      style={defaultStyle(true)}
+    >
+      <div
```
```suggestion
<span className='sr-only'>
        {t('learn.percent-complete', {
          percent: completedPercent
        })}
      </span>
      <div
```

bbsmooth commented:
I also think it would be best if we add the percentage completed to the end of the congrats message using sr-only text since it is dynamically added content that screen users might be interested in knowing as well (without having to traverse the DOM to find it). In addition, I think we can get rid of the special keyboard instructions for screen readers we are currently adding to the end of the congrats message. Screen reader users don't need to hear it over and over again. 

Would it be OK if I pushed these changes to lower-jaw.tsx?

Ismailtlem commented:
Hi,

- I removed the disabled es-lint rules in the `show` component'
- Replaced the inline style 
- Fixed the completion modal body test
- Added the course name on top of the progress bar. Let me know @ahmadabdolsaheb if that's ok or you want to change that
 
![Screenshot from 2022-10-14 13-11-49](https://user-images.githubusercontent.com/34961373/195843933-3b3b469c-52b0-4bfb-bc00-74c01b5e89e4.png)

I'll continue on the other points

ahmadabdolsaheb commented:
Thanks @Ismailtlem. Once you are done, I will go in and play with the styles and see what to do with the course name.

Ismailtlem commented:
'Depending on where I am scrolled to, the lower jaw is sometimes off screen '

- Can someone provide a screenshot of this ? @moT01  
I don't know how to reproduce it

Ismailtlem commented:
Hi @ahmadabdolsaheb 

'If we are only using the current block Ids to calculate the completedPercent, you should query currentBlockIds as high up the component hierarchy, calculate completedPercent, and pass completedPercent props to children. That would simplify the flow of information. If show is the parent component for two of the consumers of the completedPercent, let's calculate completedPercent there once and pass it to MultifileEditor and completion-modal.'

- Is it not the case that the completedPercent is calculated in the `show` component ?

ahmadabdolsaheb commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -225,135 +208,79 @@ export class CompletionModalInner extends Component<
-};
-
 const CompletionModal = (props: CompletionModalsProps) => {
   const currentBlockIds = useCurrentBlockIds(
```
> Is it not the case that the completedPercent is calculated in the show component ?
blockIds are being queried here to set the state of parent component. That should not be necessary. 

The BlockIds are queried here and set as state on the parent component using setCurrentBlockIds.

If BlockIds are needed on the parent component, query them there and pass only what is necessary to children components.

ahmaxed reviewed (commented):
Thanks for implementing some of the requested changes.

Here are a few nitpicks. Let me know if it makes sense and if you need further clarification.

moT01 commented:
> 'Depending on where I am scrolled to, the lower jaw is sometimes off screen '
>
>    Can someone provide a screenshot of this ? @moT01
>    I don't know how to reproduce it

You probably don't need to do anything about that here @Ismailtlem. To reproduce, just scroll the lower jaw off the screen and use the hotkeys to run the tests. You will pass the step, but can't really tell because the lower jaw is not on the screen. I would probably save this for another PR. I suppose the solution would be to auto scroll that into view.

Ismailtlem commented:
Commits here got a bit messy, maybe we should make some cleaner pr once the review is finished

raisedadead commented:
> Commits here got a bit messy, maybe I can make some cleaner pr once the review is finished

Do not worry about it, we squash commits on merge – meaning as long as the changes on the files tab is what you want, you are good.

moT01 commented on client/src/templates/Challenges/components/completion-modal-body.test.tsx:
```diff
@@ -32,14 +32,11 @@ describe('<CompletionModalBody />', () => {
-      expect(screen.getByTestId('fcc-progress-bar-percent')).toHaveStyle({
-        width: `${props.completedPercent}%`
-      });
+      setTimeout(() => {
```
It doesn't seem like we would need it - there's no delay so it doesn't do anything I presume.

Ismailtlem replied:
removed

moT01 commented on client/src/components/ProgressBar/progress-bar.tsx:
```diff
@@ -0,0 +1,81 @@
+
+function ProgressBar({ completedPercent }: ProgressBarProps): JSX.Element {
+  const [shownPercent, setShownPercent] = useState(0);
+  const [progressInterval, setProgressIntervall] = useState(0);
```
```suggestion
  const [progressInterval, setProgressInterval] = useState(0);
```

moT01 commented on client/src/components/ProgressBar/progress-bar.tsx:
```diff
@@ -0,0 +1,81 @@
+      if (percent >= completedPercent) clearInterval(myInterval);
+    }, intervalLength);
+
+    setProgressIntervall(myInterval);
```
minor typo
```suggestion
    setProgressInterval(myInterval);
```

Ismailtlem replied:
Corrected

moT01 reviewed (changes requested):
I gave this some more testing. I found one issue that we for sure need to fix here - see comment below about the text not showing up.

I found a second issue that seemed kind of strange. When I submit a front end libraries project, I see 0 of 0 projects completed:
![Screen Shot 2022-10-31 at 1 37 33 PM](https://user-images.githubusercontent.com/20648924/199084149-8b4ac040-444a-4d51-ab1f-0ebd3cc0f824.png)
^ The percent under there is 100% whether I've done 1, 2, 3 or 4 of the projects. And when I complete all 5, I get NaN:
<img width="433" alt="Screen Shot 2022-10-31 at 1 36 19 PM" src="https://user-images.githubusercontent.com/20648924/199083920-863c679a-6ab7-459d-81b7-d697232360a2.png">
This only seemed to happen with the front-end-libs projects block. I didn't test all the other blocks, but I didn't see it on the few others I tested. It doesn't happen on .dev, so it seems like it might be something on this PR? not sure. I suppose we should probably look into it here, as well.

A couple side thoughts that we don't need to change here at the moment:
I kinda feel like the lower jaw should be in this order:
-Congrats message
-Progress Bar
-Submit button
-Other buttons (reset)

And I'm wonder if we should put the block title in the progress bar (wrap the text when needed), making these two suggestions look kinda like this:
![Screen Shot 2022-10-31 at 12 37 30 PM](https://user-images.githubusercontent.com/20648924/199082745-322e62a3-b92b-44ed-be18-ce5c967b4d61.png)

Ismailtlem commented:
- Made the order as the following : Congrats message, Progress Bar, Submit button, Other buttons (reset). There might be some additional fixes to do
- A bit stuck in how to make the foreground text appear in the progress bar, looks easy but I don't know what I am missing. Any idea ?
- Will investigate the issue you raised @moT01 when I have some free time and let you know

ahmadabdolsaheb commented:
@Ismailtlem, let me know when you have addressed the issues that Tom has raised. 
I will get in and sort out the styles.

Ismailtlem commented:
Hi,

Sorry for my late replies on this PR. 

Corrected the first issue raised by @moT01 regarding the incorrect frontend project total numbers. We had to get the currentBlockIds in the places where `<CompletionModal />` was used. The rules-of-hooks is disabled because react doesn't allow using hooks in class component. I guess there is a better solution than this one, maybe refactoring those class component into functional components but we can do it in another PR.

There is still some styling to fix here.

ahmadabdolsaheb commented:
> Hi,
> 
> Sorry for my late replies on this PR.
> 
> Corrected the first issue raised by @moT01 regarding the incorrect frontend project total numbers. We had to get the currentBlockIds in the places where `<CompletionModal />` was used. The `rules-of-hooks` is disabled because react doesn't allow using hooks in class component. I guess there is a better solution than this one, maybe refactoring those class component into functional components but we can do it in another PR.
> 
> There is still some styling to fix here.


What are the issues that you came across?

raisedadead commented:
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

raisedadead commented:
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

Ismailtlem commented:
The issue I had @ahmadabdolsaheb was the progress bar foreground style, but I made a fix

ahmadabdolsaheb commented:
@mot01, I finally got some time to review this pr.

There are two main queries that are used to calculate the progress percentage:
1) One used in the Show components which pulls up the related challenge info
2) Another previously used in completion-modal (now used in the show components) which pulls up all challenges data.

There is some room for optimization here:
1) The second query should be kept close to the progress-bar so it only pulls the data when necessary.
2) Use dynamic queries to pull only the related data on the second query
3) Since the some current challenge data (first query) are needed to calculate the percentages, we should pass them downward to where the second query is executed.
4) A progress bar wrapper component could be used to receive the data from the first query, run the second query, calculate percentage, and pass it to the progress-bar component.

Other to dos:
- styles need to be adjusted for the progress bar so it is optimized for the modal and lower jaw.

If this sounds good to you, I could proceed with the modifications.

ahmadabdolsaheb commented:
It turned out the progress bar could be an isolated component and would not need to receive any props from its parent.

Since the challengeMetaSelector provides all challenge related data, progress-bar component could query for the currentBlockIds and calculate completedPercent.

Here is the example: #49110

The main issue is that allowBlockDonationRequests in completion modal needs to know the completedPercent in order pop the block donation modal.

ahmadabdolsaheb commented:
@moT01 could you give this another round of reviews. 
We could tackle the following optimizations in upcoming prs:
1) rewire allowBlockDonationRequests
2) create a progressbar wrapper that pulls challenge data from challengeMetaSelector and handles all progressbar logic.

moT01 reviewed (commented):
This looks real good @ahmadabdolsaheb. I found two things to possibly address.

- ~~When I complete a challenge in the multifile editor, screen readers don't mention the block. e.g. "Learn HTML by Building a Cat Photo App, 10% complete" - it just says "10% complete". We maybe want to add that for screen readers.~~ Edit: This was intentional and we can leave it as-is

- I'm getting some weird behavior in the multifile editor. To reproduce, go to a multifile challenge with the lower jaw, complete the challenge but don't submit, highlight the editor and press tab:
![Feb-01-2023 12-26-49](https://user-images.githubusercontent.com/20648924/216139461-66091db4-2acf-4b8a-b276-c94bf6ce8937.gif)

The percent goes from 10% (where I'm supposed to be), to 100%. I was able to reproduce pretty consistently, but it didn't always happen. I dug into it a bit, couldn't quite figure it out - let me know if you want to get on a call and see what the problem is.

bbsmooth commented:
> When I complete a challenge in the multifile editor, screen readers don't mention the block. e.g. "Learn HTML by Building a Cat Photo App, 10% complete" - it just says "10% complete". We maybe want to add that for screen readers.

I think I was the one who suggested we announce only the percentage to screen reader users instead of announcing both course name and percentage. The reasoning being that hearing the course name every time is probably unnecessary because the user knows what course they are in, and so it just becomes tedious repetition.

ahmadabdolsaheb commented on cypress/e2e/default/learn/challenges/projects.js:
```diff
@@ -185,3 +185,31 @@ describe('project submission', () => {
+        '/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-1'
+      );
+      cy.contains('button', 'Start coding!').click({ force: true });
+      cy.get(`${selectors.editor} textarea`, { timeout: 16000 })
```
There are similar timeouts in the repo when accessing the editor. So I based this on those.

ahmadabdolsaheb commented:
@Ismailtlem, could you address Tom's concern and any failing tests if possible.

ahmadabdolsaheb commented:
I am pushing the query data to redux so we could pull them from anywhere.

BalanaguYashwanth commented:
Hello @moT01 ,

Please review this pull request, so that upcoming features like share button to build below the progress bar, for those it won't become blockers

Sboonny commented:
@BalanaguYashwanth, we don't need a project manager, especially one that ask people to look at something at their weekend.

Thank you for your excitement, and your every feedback on issue and PR so far.

ahmadabdolsaheb commented:
This pr should go in right after https://github.com/freeCodeCamp/freeCodeCamp/pull/49381.

Sembauke commented:
@Ismailtlem Sorry, for the inconvinence. I have opened a new pr in your name with the required changes to add the rest of the progress bar #50167 .
